Include a link in PyNEST examples to run them as notebooks

jessica-mitchell opened the pull request:
This PR  adds a link (see [nbgitpuller](https://hub.jupyter.org/nbgitpuller/link.html)) to the PyNEST examples so they can run as a  notebook - using the  EBRAINS JupyterHub as service.

The link is added at build time so it is not in the source repository.
It is added to the reStructuredText files in `auto_examples`  generated by sphinx_gallery.
The notebooks are located in nest/nest-simulator-examples/ repository.
This PR also includes a file that explains how to run the notebooks and some basic troubleshooting.
The Python download button generated by sphinx_gallery remains but the notebook button is removed.

This PR only deals with the sphinx-build of the link. 
The notebooks are already available but the automated generation/maintenance of the notebooks will be dealt with in another PR.

See sample: https://nest-test.readthedocs.io/en/add-link-examples/auto_examples/one_neuron.html

janskaar commented on doc/htmldoc/conf.py:
```diff
@@ -247,6 +248,81 @@ def config_inited_handler(app, config):
+
+    # Find all relevant files
+    # Inject prolog into Python example
+    files = list(Path(doc_build_dir / "auto_examples/").rglob("**/*.rst"))
```
```suggestion
    files = list(Path(doc_build_dir / "auto_examples/").rglob(*.rst"))
```
Don't think we need the ** in rglob, only in glob?

jessica-mitchell replied:
Yes you're right I removed the ** and reran the examples and they seem to work as expected.

janskaar reviewed (commented):
The documentation builds for me locally, and everything looks the same as in your example, and the links to the ebrains notebooks work. The user guide looks great.

I'm getting a few warnings when building the documentation that are not present when I build from the master branch: 

`/home/janeirik/nest_repository/2605/nest-simulator-build/doc/htmldoc/auto_examples/compartmental_model/receptors_and_current.rst:34: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/janeirik/nest_repository/2605/nest-simulator-build/doc/htmldoc/auto_examples/compartmental_model/two_comps.rst:34: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/janeirik/nest_repository/2605/nest-simulator-build/doc/htmldoc/auto_examples/pong/generate_gif.rst:34: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/janeirik/nest_repository/2605/nest-simulator-build/doc/htmldoc/auto_examples/pong/networks.rst:34: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/janeirik/nest_repository/2605/nest-simulator-build/doc/htmldoc/auto_examples/pong/pong.rst:34: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/janeirik/nest_repository/2605/nest-simulator-build/doc/htmldoc/auto_examples/pong/run_simulations.rst:34: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/janeirik/nest_repository/2605/nest-simulator-build/doc/htmldoc/auto_examples/sudoku/plot_progress.rst:34: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/janeirik/nest_repository/2605/nest-simulator-build/doc/htmldoc/auto_examples/sudoku/sudoku_net.rst:34: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/janeirik/nest_repository/2605/nest-simulator-build/doc/htmldoc/faqs`

I'm also getting some warnings that are also present when building it from the master branch, mostly related to HillTononiModels.ipynb and hl_api_*.py.

jessica-mitchell commented on doc/htmldoc/conf.py:
```diff
@@ -247,6 +248,81 @@ def config_inited_handler(app, config):
+     The link to run the notebook is rendered in an image within a card directive.
+    """
+    example_prolog = """
+.. only:: html
```
Yes, exactly

jessica-mitchell commented on doc/htmldoc/conf.py:
```diff
@@ -247,6 +248,81 @@ def config_inited_handler(app, config):
+        get_path = file.parent
+        get_subdir = get_path.stem
+
+        lines = []
```
removed

jessica-mitchell commented:
@janskaar Thanks for the review, I addressed your comments. As for the sphinx warnings you see at build time, we are aware of them and you can ignore them as they need to be dealt with in separate issues.

jessica-mitchell commented:
@med-ayssar @janskaar Can you please take another look?  thanks!

med-ayssar commented on doc/htmldoc/conf.py:
```diff
@@ -247,6 +248,80 @@ def config_inited_handler(app, config):
+        # insert prolog into rst file after heading
+        lines.insert(i + 1, prolog + '\n')
+
+        with open(file, 'w') as f:
+            lines = "".join(lines)
+            f.write(lines)
+
```
We don't need to open the file twice: once for reading and once for writing: I would do the following:

```python
with open(file, "r+") as f:
    data = f.readlines()
    data.modify()  # modify the line you want
    data = "".join(data)
    f.truncate(0) # clear everything from the file
    f.seek(0) # set pointer to the beginning of the file
    f.write(data)
```

med-ayssar replied:
@Helveg, @jougs & @terhorstd: Any opinions regarding the use of `truncate` and `seek` instead of opening the file twice?

Helveg replied:
I'd say it's premature optimization at best. Are we on the clock? :)

med-ayssar replied:
Then @jessica-mitchell is up to decide which variant to use :p

jessica-mitchell replied:
I went with back with the previous read/ write since it is what I'm familiar with.

med-ayssar reviewed (commented):
Otherwise, the rest looks good

med-ayssar commented on doc/htmldoc/conf.py:
```diff
@@ -247,6 +248,80 @@ def config_inited_handler(app, config):
+
+            # get name of file and subdirectory to replace in link target
+            if get_subdir != "auto_examples":
+                path2example = get_subdir + "%2F" + file.stem
```
Then my subsequent suggestion will solve this already and handle all levels.

jessica-mitchell commented:
@med-ayssar I made the changes you suggested

med-ayssar commented on doc/htmldoc/conf.py:
```diff
@@ -293,34 +293,25 @@ def add_button_to_examples(app, env, docnames):
-            lines = f.readlines()
-
+        with open(file, "r+") as f:
+            data = f.readlines()
```
Maybe read the lines at the bottom

jessica-mitchell replied:
sorry what do you mean?

jessica-mitchell replied:
I just moved it a few lines down, which I think is what you meant?

med-ayssar replied:
I mean move opening of the file and the reading to the bottom after computing the `prolog` variable. Because you don't need to open the file before that.

med-ayssar replied:
It looks better now, thank you.

med-ayssar commented:
Could you post the new `link` to see it works fine with these new changes?

jessica-mitchell commented:
new link: https://nest-test.readthedocs.io/en/add-link-examples/auto_examples/lin_rate_ipn_network.html

jessica-mitchell commented:
@med-ayssar I fixed the link by adding `splitext` line.

https://nest-test.readthedocs.io/en/add-link-examples/auto_examples/balancedneuron.html

heplesser commented:
@med-ayssar @janskaar Could you please resolve all conversations that you opened so that we can merge? Our general policy is that the person who started a conversation should mark it as resolved.

janskaar commented:
> @med-ayssar @janskaar Could you please resolve all conversations that you opened so that we can merge? Our general policy is that the person who started a conversation should mark it as resolved.

med-ayssar commented:
Done!

janskaar commented:
I can't find any resolve button for my conversations. Is there a special
permission or something needed that I don't have?

On Tue, 21 Mar 2023, 09:25 Hans Ekkehard Plesser, ***@***.***>
wrote:

> @med-ayssar <https://github.com/med-ayssar> @janskaar
> <https://github.com/janskaar> Could you please resolve all conversations
> that you opened so that we can merge? Our general policy is that the person
> who started a conversation should mark it as resolved.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nest/nest-simulator/pull/2605#issuecomment-1477435175>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHACQZN76KHMTZNYPS6AQOTW5FQXVANCNFSM6AAAAAAUSPZ5ZI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

med-ayssar commented:
@janskaar I think you forgot to post the photo, but if you don't mind, I can resolve your threads instead.

janskaar commented:
Yes, please do!

On Tue, 21 Mar 2023, 14:16 med-ayssar, ***@***.***> wrote:

> @janskaar <https://github.com/janskaar> I think you forgot to post the
> photo, but if you don't mind, I can resolve your threads instead.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nest/nest-simulator/pull/2605#issuecomment-1477820362>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHACQZLO4FSRET32AZS5A7DW5GSZHANCNFSM6AAAAAAUSPZ5ZI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
